### ISSUE TITLE
False positive `Substratum` corpses

### ISSUE THREAD

**Issue by d4rken** [d4rken/sdmaid-public#1813]
> In Oreo ROMs, When selecting /data/app in Corpus Finder options, SD Maid identifies Substratum overlays as Corpus to be removed. That was not the case in Nougat ROM.
I disabled this option for now.

![screenshot_20180702-125621_sd maid](https://user-images.githubusercontent.com/1439229/42281369-9d5f87b4-7fa3-11e8-9468-a92e18209169.jpg)

**Comment by d4rken** [comment]
```
1530645775512 V/RXS:Harvester:Output: 644:regular file:66306:1308300:1000:1000:8520:512:24:1530537270:`/data/app/com.amazon.mp3.SwiftBlack.Android8Samsung-9KvWnlV6l60TcuWCKl1hLw==/base.apk'
1530645775512 D/RXS:Harvester:Output: 644:regular file:66306:1308300:1000:1000:8520:512:24:1530537270:`/data/app/com.amazon.mp3.SwiftBlack.Android8Samsung-9KvWnlV6l60TcuWCKl1hLw==/base.apk'
1530645775512 V/RXS:Harvester:Output: 755:directory:66306:1308301:1000:1000:4096:512:8:1530537270:`/data/app/com.amazon.mp3.SwiftBlack.Android8Samsung-9KvWnlV6l60TcuWCKl1hLw==/lib'
1530645775513 D/RXS:Harvester:Output: 755:directory:66306:1308301:1000:1000:4096:512:8:1530537270:`/data/app/com.amazon.mp3.SwiftBlack.Android8Samsung-9KvWnlV6l60TcuWCKl1hLw==/lib'
```

**Comment by yelkarama** [comment]
Here are suggested steps to reproduce the issue:
1- Install substratum app
2- Install a substratum theme app (I'm using Swift Black which is paid but I guess a free one can do too)
3- Open substratum app and select the installed theme
4- Select some applications and choose "Build and enable" action. Give it some time until seeing "Up to date" status on all the selected apps
5- reboot the device
6- After the boot completed, open the themed apps and verify the substratum theme was applied
7- Open SDmaid, ensure /data/app option is enabled in Corpus Finder settings
8- Run the Corpus finder. It should come back with entries for all themed apps
9- Clean all these Corpuses
10- Open the themed apps and verify the substratum theme was removed

**Comment by yelkarama** [comment]
On my samsung S7, I found this hidden xml file created by substratum, that lists all the installed overlays:
/sdcard/.samsung_overlays.xml
[samsung_overlays.xml.txt](https://github.com/d4rken/sdmaid-public/files/2163944/samsung_overlays.xml.txt)
(added .txt extension to be able to upload it)

**Comment by d4rken** [comment]
> On my samsung S7, I found this hidden xml file created by substratum, that lists all the installed overlays:
/sdcard/.samsung_overlays.xml

Looks interesting, might be the solution, though on first glance it just's lists the package names of the "themed" originals. We'd need something that gives us the relation between original-pkg<>themed-pkg.

Just knowing the themed package can help prevent false positives, but then we wouldn't be able to list an old-theme as corpse if the original app is uninstalled.

**Comment by SuperSandro2000** [comment]
> but then we wouldn't be able to list an old-theme as corpse if the original app is uninstalled.

This is not required because substratum handles this by itself. I tested it and it left 0 chunk behind.
My steps:
1. Installed overlay for an App
2. rebooted (tested with Sungstratum root mode)
3. tested the overlay
4. removed the app, substratum removed the overlay
5. rebooted and substratum removed the overlay apk in /sdcard/.substratum

I don't think there is any cleaning work required.

**Comment by d4rken** [comment]
> This is not required because substratum handles this by itself. I tested it and it left 0 chunk behind.

Yes but that isn't a satisfactory solution for SD Maid, we want SD Maid to clean up when other apps fuck up. So ultimatively each apps fate in terms of being a corpse or not should be tied to the package of an installed app, not a file on the sdcard.

What happens if substratum looses track of the "themed" apps? Does it have a way to build it's database of themed apps again or is the user left with the files until a factory reset?

> Isn't there any option to hide by default all overlays of any nature?

That's the whole issue we have here 😁 . How do we detect them?

I think I'll have to setup a test device and take a closer look at the "theme" directories. It looks like there was a `base.apk` in them, maybe we can get an original packagename from that.

**Comment by SuperSandro2000** [comment]
> So ultimatively each apps fate in terms of being a corpse or not should be tied to the package of an installed app, not a file on the sdcard.

I deleted the xml file on the SD card and substratum still knows his overlays. Iirc it is only used to activate overlays and deactivate overlays. Right now it is out of sync for me and still everything works.
The only way you can end up with overlays installed is if you uninstall substratum.

> What happens if substratum looses track of the "themed" apps? Does it have a way to build it's database of themed apps again or is the user left with the files until a factory reset?
Then you need to do it on purpose. The only way would be if SD maid ignores storage management (what it really shouldn't) or a user clears the data by doing into the root folder and deleting it.

> How do we detect them?
All overlays have the same permissions at least for me on an samsung device. Screenshot attached.

The android manifest in the theme apk stores the parent theme. I need to take a closer look of I can find the target app.
The package name of the theme apk is always extended from the themed app. Maybe this helps.
Keep in mind those apk are not real apps and aren't installed in a normal fashion.

**Comment by d4rken** [comment]
> The android manifest in the theme apk stores the parent theme. I need to take a closer look of I can find the target app.

If SD Maid can lookup it up the same way that sounds like the best solution so far.

> The package name of the theme apk is always extended from the themed app. Maybe this helps.

Similar, but afaik not always the same, might be a bit fragile to try and match this especially as it would require SD Maid to know all theme names that exist?

> All overlays have the same permissions at least for me on an samsung device. Screenshot attached.

The samsung "overlay" permission looks unique enough, but then we end up with the same issue of basically hardcoding exclusions instead of improving detection. Do other ROMs support theming but don't use this samsung overlay permission?

Looking up the package name via the apk sounds good so far (if it is possible), we could additionally filter out apps by checking for such specific permissions to improve scan performance, depending on whether this is reliable enough.

After "theming" an app, is it required to keep substratum installed for it to work?

**Comment by SuperSandro2000** [comment]
> If SD Maid can lookup it up the same way that sounds like the best solution so far.

````
    <uses-permission android:name="com.samsung.android.permission.SAMSUNG_OVERLAY_COMPONENT"/>
    <overlay android:priority="1" android:targetPackage="com.google.android.apps.maps"/>
    <application android:allowBackup="false" android:hasCode="false">
        <meta-data android:name="app_version" android:value="v=9.81.1"/>
        <meta-data android:name="app_version_code" android:value="981100244"/>
        <meta-data android:name="overlay_version" android:value="2"/>
        <meta-data android:name="target_package" android:value="com.google.android.apps.maps"/>
    </application>
````
overlay_version, target_package and Samsung_overlay_component should be very helpfull.

> Do other ROMs support theming but don't use this samsung overlay permission?

basically every ROM with RRO, Interfacer (deprecated), masquerade (N only IIRC), Marshmallow or Oreo support theming in theory. Other theming solution could be different and/or even substratum could ditch that permission on non samsung devices. Sadly I don't own one and can't test this. (To lazy to setup an emulator 😛 )

> After "theming" an app, is it required to keep substratum installed for it to work?

Did a quick test and does not seem to be necessary. if you uninstall the theme original apk then all overlays are removed too. Don't know for Swift Installer but don't want to rebuild all overlays right now.

**Comment by d4rken** [comment]
> basically every ROM with RRO, Interfacer (deprecated), masquerade (N only IIRC), Marshmallow or Oreo support theming in theory. 

Do we know if all these theming approaches use `target_package` ?

If you or @yelkarama would like to help me test an early build before an official beta, shoot me a quick mail 😄 .

**Comment by d4rken** [comment]
> TLDR:
> I would suggest to blacklist all packages that use this permission com.samsung.android.permission.SAMSUNG_OVERLAY_*, have the overlay_version or target_package meta data or blacklist them individual.

We could do a priority list:

1. Other checks failed and we are currently considering the item a corpse
2. Does the folder contain a `base.apk`? No->Corpse, Yes->Continue
3. Does the APK have a `target_package` field? No->Continue,Yes->Corpse(depending on package install status)
4. Does the APK contain  `overlay_version` or `com.samsung.android.permission.SAMSUNG_OVERLAY_*`? Yes->Not a corpse, No->Corpse


>Maybe it is worth to put them into a new cateorgy called overlays?

You mean a new corpse filter? Hmmm normally they are grouped by location not "type" as otherwise it becomes more difficult to determine which filter is responsible for what.

**Comment by SuperSandro2000** [comment]
Sounds like a plan. Would love to test it.

Idk depends on what you want for your app

**Comment by d4rken** [comment]
> Would love to test it.

I missjudged my progress a bit 😁 . Got held up with the CorpseFinder UI code, currently finishing the AppCleaner UI code, I hope to have something by tomorrow evening, maybe monday.

**Comment by AaronPeepster** [comment]
Please take me off of the email list please.

**Comment by d4rken** [comment]
I can't reproduce this so far.
I've installed this custom ROM on a Moto G5 (2015): https://downloads.aospextended.com/osprey/

Any themed app actually gets its own installed package and is thus not detected as corpse by SD Maid.

Is this issue a _Samsung Thing ™️_  ?

**Comment by SuperSandro2000** [comment]
I tested it with the only available AOSP ROM for my klte and it seems to be a Samsung thingy.

**Comment by d4rken** [comment]
Hm, I just flashed a Nexus5 with an older version such that substratum uses "RRO2" as theme engine and the issue doesn't occur either. Samsung 😢 .

I don't have a recent Samsung device at hand. Can someone zip and send me a few of the folders that are detected as corpses?

**Comment by SuperSandro2000** [comment]
A Samsung store theme (starts with theme), a substratum theme (the whatsapp one) and one from swift installer (has swift in the folder name).
[theme cropses.zip](https://github.com/d4rken/sdmaid-public/files/2175971/theme.cropses.zip)

**Comment by d4rken** [comment]
@yelkarama and @SuperSandro2000 confirmed the changes to be working. 

This also closes #1815 

@gagabs if you would like to do a preview test of the changes, send me a short mail.

Thanks for everyones help, this is the fruit of hour labor 👍 :

```java
private boolean checkApkName(OwnerInfo ownerInfo, String name) {
    SDMFile apk = JavaFile.build(ownerInfo.getItem(), name);
    SDMPkgInfo pkgInfo = getIPCFunnel().submit(new IPCFunnel.ArchiveQuery(apk, 0));
    if (pkgInfo == null) return false;

    // Normal app :')
    if (isInstalled(pkgInfo.getPackageName())) {
        ownerInfo.addOwner(new Owner(pkgInfo.getPackageName()));
        return true;
    }

    // It might be theme/overlay...
    // https://github.com/d4rken/sdmaid-public/issues/1813

    // <overlay android:priority="1" android:targetPackage="android"/>
    if (ApiHelper.hasLolliPop()) {
        String targetPkg = null;
        try {
            targetPkg = pkgInfo.tryField("overlayTarget");
        } catch (Exception e) {
            Timber.tag(TAG).w(e);
        }

        if (targetPkg != null) {
            Timber.tag(TAG).d("Target via reflection, PackageInfo.overlayTarget=%s from %s", targetPkg, apk);
            if (isInstalled(targetPkg)) {
                ownerInfo.addOwner(new Owner(targetPkg));
                return true;
            }
        }
    }

    pkgInfo = getIPCFunnel().submit(new IPCFunnel.ArchiveQuery(apk, PackageManager.GET_META_DATA | PackageManager.GET_PERMISSIONS));
    if (pkgInfo == null) return false;

    // <meta-data android:name="target_package" android:value="android"/>
    if (pkgInfo.getApplicationInfo() != null && pkgInfo.getApplicationInfo().metaData != null) {
        String targetPkg = pkgInfo.getApplicationInfo().metaData.getString("target_package");
        if (targetPkg != null) {
            Timber.tag(TAG).d("Target via metadata, target_package=%s from %s", targetPkg, apk);
            if (isInstalled(targetPkg)) {
                ownerInfo.addOwner(new Owner(targetPkg));
                return true;
            }
        }
    }

    // <meta-data android:name="Substratum_Target" android:value="com.whatsapp"/>
    if (pkgInfo.getApplicationInfo() != null && pkgInfo.getApplicationInfo().metaData != null) {
        String targetPkg = pkgInfo.getApplicationInfo().metaData.getString("Substratum_Target");
        if (targetPkg != null) {
            Timber.tag(TAG).d("Target via metadata, Substratum_Target=%s from %s", targetPkg, apk);
            if (isInstalled(targetPkg)) {
                ownerInfo.addOwner(new Owner(targetPkg));
                return true;
            }
        }
    }

    // <uses-permission d1p1:name="com.samsung.android.permission.SAMSUNG_OVERLAY_APPICON"/>
    if (pkgInfo.getRequestedPermissions() != null) {
        boolean justSamsungThings = false;
        for (String p : pkgInfo.getRequestedPermissions()) {
            if (p.contains("com.samsung.android.permission.SAMSUNG_OVERLAY_")) {
                Timber.tag(TAG).d("Unknown overlay with SAMSUNG overlay permission: %s", apk);
                justSamsungThings = true;
                break;
            }
        }
        if (justSamsungThings) {
            ownerInfo.setUnknownOwner(true);
            return true;
        }
    }

    return false;
}
